Feature - user subscribe publisher.

DeemaP opened the pull request:
add AppConfig file, combined ExecutorConfig.java, OpenApiConfig.java, ObjectMapper. add tests for publisher
refactor SubscribeService validator

DeemaP committed:
Feature - user subscribe publisher.

add AppConfig file, combined ExecutorConfig.java, OpenApiConfig.java, ObjectMapper.
add tests for publisher

drewbal commented on src/main/java/school/faang/user_service/config/redis/RedisProperties.java:
```diff
@@ -1,5 +1,6 @@
 package school.faang.user_service.config.redis;
 
+import com.fasterxml.jackson.annotation.JsonProperty;
```
лишнее уже, не?

DeemaP replied:
Да, забыл потереть)

drewbal commented on src/main/java/school/faang/user_service/controller/SubscriptionController.java:
```diff
@@ -27,7 +29,9 @@ public class SubscriptionController {
 
     @PostMapping("/{followerId}/follow/{followeeId}")
     public ResponseEntity<Void> followUser(@Valid @PathVariable long followerId, @Valid @PathVariable long followeeId) {
+        log.info("Received request to subscribe user #{} to user #{}.", followerId, followeeId);
```
Received request from user id: {} to subscribe to user id: {}

И без #

drewbal commented on src/main/java/school/faang/user_service/publisher/RecommendationReceivedEventPublisher.java:
```diff
@@ -29,6 +29,6 @@ public class RecommendationReceivedEventPublisher implements EventPublisher<Reco
     public void publish(RecommendationReceivedEvent event) {
         redisTemplate.convertAndSend(redisProperties.getChannel().getRecommendationChannel(), event);
-        log.info("Message sent to channel: {}", redisProperties.getChannel().getRecommendationChannel());
+        log.info("Event sent to channel: {}", redisProperties.getChannel().getRecommendationChannel());
```
New Recommendation event sent to channel id: {}

drewbal commented on src/main/java/school/faang/user_service/publisher/SubscriptionEventPublisher.java:
```diff
@@ -0,0 +1,34 @@
+    )
+    public void publish(SubscriptionEvent message) {
+        redisTemplate.convertAndSend(redisProperties.getChannel().getSubscriptionChannel(), message);
+        log.info("Event sent to channel: {}", redisProperties.getChannel().getRecommendationChannel());
```
New Subscription event sent to channel id: {}

drewbal commented on src/main/java/school/faang/user_service/service/MentorshipRequestService.java:
```diff
@@ -93,5 +93,4 @@ private MentorshipRequest validateAndGetMentorshipRequest(long id) {
             return new EntityNotFoundException("Request with id #" + id + " not found.");
```
без #

drewbal commented on src/main/java/school/faang/user_service/service/SubscriptionService.java:
```diff
@@ -1,45 +1,61 @@
+                followerId, followeeId);
+
+        subscriptionEventPublisher.publish(new SubscriptionEvent(followerId, followeeId, subscribedAt));
+        log.info("User #{} successfully subscribed to user #{}.", followerId, followeeId);
```
без #

drewbal commented on src/main/java/school/faang/user_service/validator/SubscriptionValidator.java:
```diff
@@ -11,21 +11,26 @@ public class SubscriptionValidator {
         if (subscriptionRepository.existsByFollowerIdAndFolloweeId(followerId, followeeId)) {
-            throw new DataValidationException("Subscription already exists.");
+            throw new DataValidationException(
+                    String.format("User #%d is already subscribed to user #%d.",
```
без #

drewbal commented on src/main/java/school/faang/user_service/validator/SubscriptionValidator.java:
```diff
@@ -11,21 +11,26 @@ public class SubscriptionValidator {
         if (!subscriptionRepository.existsByFollowerIdAndFolloweeId(followerId, followeeId)) {
-            throw new DataValidationException("Subscription does not exist.");
+            throw new DataValidationException(
+                    String.format("User #%d is not subscribed to user #%d.",
```
без #

drewbal commented on src/main/java/school/faang/user_service/validator/SubscriptionValidator.java:
```diff
@@ -11,21 +11,26 @@ public class SubscriptionValidator {
+    public void validateNoSelfSubscription(long followerId, long followeeId) {
+        if (followerId == followeeId) {
+            throw new DataValidationException(
+                    String.format("User #%d cannot subscribe to themselves.", followerId));
```
Users, раз themselves
без #

drewbal commented on src/test/java/school/faang/user_service/service/SubscriptionServiceTest.java:
```diff
@@ -51,7 +69,7 @@ void getFollowersCount_shouldReturnCount() {
         assertEquals(5, count);
 
-        verify(subscriptionValidator).validateUserExists(followeeId);
+        verify(userValidator).validateUserById(followeeId);
```
можно times добавить для единообразия подхода

drewbal commented on src/test/java/school/faang/user_service/service/SubscriptionServiceTest.java:
```diff
@@ -61,6 +79,6 @@ void getFollowingCount_shouldReturnCount() {
         assertEquals(3, count);
 
-        verify(subscriptionValidator).validateUserExists(followerId);
+        verify(userValidator).validateUserById(followerId);
```
можно times добавить для единообразия подхода

drewbal reviewed (commented):
фигульки

DeemaP committed:
Merge branch 'refs/heads/Dragon-master-stream7' into feature/bjs2-41379-user-subscribes-analytics

# Conflicts:
#	src/main/java/school/faang/user_service/config/redis/RedisConfig.java

DeemaP committed:
Resolve merge conflicts.
Apply changes based on cross-review comments.

RuslanZinovyev commented on src/main/java/school/faang/user_service/controller/SubscriptionController.java:
```diff
@@ -27,7 +29,9 @@ public class SubscriptionController {
 
     @PostMapping("/{followerId}/follow/{followeeId}")
```
Я понимаю что это старый код, но в ендпоинтах лучше придерживаться ресурсов, а не действий. 
`POST /users/{followerId}/followers`
А идентификатор можно в теле передать через DTO 
{ "followeeId": 12345 }

DeemaP replied:
Спасибо

Я немного переделал, чтоб читалось понятно

А есть смысл ради одного параметра делать целую дто? Расширять тут будет нечего, тут либо follow, либо unfollow и все, добавлять будет нечего.
Я лучше тогда followeeId кину в requestParam, чем лишний класс создавать.

Плюс я только обратил внимание на аннотации @Valid, которые не нужны) Тоже поправил в этом эндпоинте.

Остальные эндпоинты Витя будет поправлять сам, мы с ним договорились.

RuslanZinovyev commented on src/main/java/school/faang/user_service/service/SubscriptionService.java:
```diff
@@ -1,45 +1,61 @@
+        LocalDateTime subscribedAt = subscriptionRepository.findCreatedAtByFollowerIdAndFolloweeId(
+                followerId, followeeId);
+
+        subscriptionEventPublisher.publish(new SubscriptionEvent(followerId, followeeId, subscribedAt));
```
Потенциальная dual write проблема

DeemaP replied:
Мы все перелопатили в соответствии с тем решением, что ты предложил в PostService.

DeemaP committed:
Merge branch 'refs/heads/Dragon-master-stream7' into feature/bjs2-41379-user-subscribes-analytics

# Conflicts:
#	src/main/java/school/faang/user_service/config/redis/RedisConfig.java

DeemaP committed:
Commit message:

fix: align Outbox implementation with PostService and remove while-true loop

- Refactor Outbox processing to follow PostService patterns
- Replace while-true loop with an optimized processing approach
- Address reviewer comments for improved code structure and clarity

DeemaP committed:
Refactor
remove OutboxEventListener
add scheduler for OutboxProcessor triggerProcessing() every minute.

Add swagger annotations for followUser endpoint in SubscriptionController

RuslanZinovyev commented on src/main/java/school/faang/user_service/exception/GlobalExceptionHandler.java:
```diff
@@ -139,4 +139,20 @@ public ResponseEntity<String> handleRedisPublishingException(RedisPublishingExce
+    public ResponseEntity<ErrorResponse> handleJsonProcessingException(JsonProcessingException ex) {
+        log.error("Json processing exception: {}", ex.getMessage(), ex);
+        String errorMessage = extractMessage(ex.getMessage());
+        return ResponseEntity.status(HttpStatus.NOT_FOUND)
```
Тут логически скорее ближе BAD_REQUEST нежели NOT_FOUND
```suggestion
        return ResponseEntity.status(HttpStatus.BAD_REQUEST)
```

DeemaP replied:
Подумали немножко и пришли к тому, что еще логичнее INTERNAL_SERVER_ERROR, поскольку эта ошибка может возникнуть на этапе сериализации SubscriptionEvent в json, то есть внутренняя ошибка на сервере, а не проблема некорректного запроса пользователя.

DeemaP committed:
Resolve conflicts
apply changes based on reviewer comments

DeemaP committed:
Merge branch 'refs/heads/Dragon-master-stream7' into feature/bjs2-41379-user-subscribes-analytics

# Conflicts:
#	src/main/java/school/faang/user_service/config/redis/RedisProperties.java

DeemaP committed:
Resolve conflicts
Refactor RedisProperties to record
Refactor tests
